### PR TITLE
Implemented the ability to search entities.

### PR DESCRIPTION
**Pull request by Jack12816**
This update ships the implementation of the query/search functionality.
This features low-level HTTP client APIs, client driver APIs and a brand
new search package which consists of three parts. The query settings, to
frame the search indention. The paging component which takes care of
planned page layouts and request optimizations. And last but not least a
search executor which serves the Ruby Enumerable API by combining the
query settings and paging details into actual client driver actions.

## Search Architecture 

HAUSGOLD SDK - Search

## Paging

HAUSGOLD SDK - Paging

### CODE REVIEW

File: lib/hausgold/search_criteria/executor.rb [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+                if last_page? && last_page_unaligned?
+
+              # Serve the +Enumerable+ interface and yield each found entity
+              page.each { |entity| yield(entity) }
+
+              # We also stop paging when the page does not fill its boundaries,
+              # because this indicates an "unexpected" end. The planned pages
+              # MUST NOT be fully served by the API, when there is not enough
+              # data to respond. This also includes completely empty pages.
+              raise StopIteration if transmission_count < per_page
```
[the remark this answers is not in the document]
Comment (Jack12816) [reply lib/hausgold/search_criteria/executor.rb]: `page_cursor.next` raises `StopIteration`, so it's just the very same handling to signal the same case.

File: lib/hausgold/search_criteria/executor.rb [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+      extend ActiveSupport::Concern
+
+      included do
+        include Hausgold::Utils::Bangers
+        include Enumerable
+
+        # Get back the page cursor which tracks the paging execution. The
+        # cursor is an +Enumerator+ which reflects the planned pages range
+        # (+#pages+). By doing so, subsequent calls will result in the very
+        # same instance. (Memorization)
```
[the remark this answers is not in the document]
Comment (Jack12816) [reply lib/hausgold/search_criteria/executor.rb]: Yep, fixed the typo.

File: lib/hausgold/search_criteria/paging.rb [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+          first_page[:start_offset]..-1
+        end
+
+        # Just like the +#relative_first_page_result_slice+, but for the last
+        # page. The relative slicing is useful for stream processing.
+        #
+        # @return [Range] the first page relative slicing range
+        def relative_last_page_result_slice
+          # When the total is infinity, the ending is infinity, too
+          return 0..-1 if last_page[:total] == Float::INFINITY
```
Comment (marcusg) [inline lib/hausgold/search_criteria/paging.rb]: was genau hat es mit der range `0..1` auf sich?

Comment (Jack12816) [reply lib/hausgold/search_criteria/paging.rb]:
```ruby
# fiddling with float infinity is bad on +Array#slice+
[].slice(0..Float::INFINITY)
# => RangeError (float Inf out of range of integer)

# Ruby 2.6 adds endless ranges
[1, 2, 3].slice(0..)
# => [1, 2, 3]

# But we need to support Ruby down to 2.3
[1, 2, 3].slice(0..-1)
# => [1, 2, 3]
[1, 2, 3].slice(0..-2)
# => [1, 2]
```

> Array indexing starts at 0, as in C or Java. A negative index is assumed to be relative to the end of the array—that is, an index of -1 indicates the last element of the array, -2 is the next to last element in the array, and so on.

https://ruby-doc.org/core-2.2.0/Array.html

**Pull request merged**